File: backend/src/auth/auth.controller.ts
import {
  Controller,
  Post,
  Body,
  UseGuards,
  Request,
  Get,
  HttpCode,
  HttpStatus,
} from '@nestjs/common';
import { AuthService } from './auth.service';
import { RegisterDto } from './dto';
import { LocalAuthGuard, JwtAuthGuard } from './guards';

/**
 * AuthController - Handles authentication endpoints
 * 
 * Endpoints:
 * - POST /auth/register - Create new user account
 * - POST /auth/login    - Login and get JWT token
 * - GET  /auth/profile  - Get current user (protected)
 */
@Controller('auth')
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  /**
   * Register a new user
   * 
   * POST /auth/register
   * Body: { email, password, first_name, last_name, role }
   * 
   * Returns the created user (without password)
   */
  @Post('register')
  async register(@Body() dto: RegisterDto) {
    return this.authService.register(dto);
  }

  /**
   * Login with email and password
   * 
   * POST /auth/login
   * Body: { email, password }
   * 
   * The @UseGuards(LocalAuthGuard) decorator:
   * 1. Intercepts the request BEFORE this method runs
   * 2. Uses LocalStrategy to validate credentials
   * 3. If valid, attaches user to req.user
   * 4. Then this method runs and generates the JWT
   * 
   * Returns: { access_token, user }
   */
  @UseGuards(LocalAuthGuard)
  @Post('login')
  @HttpCode(HttpStatus.OK) // Return 200 instead of 201 for login
  async login(@Request() req) {
    // req.user is set by LocalAuthGuard after successful validation
    return this.authService.login(req.user);
  }

  /**
   * Get current user profile
   * 
   * GET /auth/profile
   * Header: Authorization: Bearer <token>
   * 
   * The @UseGuards(JwtAuthGuard) decorator:
   * 1. Extracts JWT from Authorization header
   * 2. Validates it using JwtStrategy
   * 3. If valid, attaches user to req.user
   * 
   * This is a test endpoint to verify JWT is working
   */
  @UseGuards(JwtAuthGuard)
  @Get('profile')
  getProfile(@Request() req) {
    return req.user;
  }
}

File: backend-java/src/main/java/com/smartsupply/controller/AuthController.java
package com.smartsupply.controller;

import com.smartsupply.dto.AuthResponse;
import com.smartsupply.dto.LoginRequest;
import com.smartsupply.dto.RegisterRequest;
import com.smartsupply.entity.User;
import com.smartsupply.service.AuthService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

/**
 * AuthController - REST endpoints for authentication.
 * 
 * Comparison:
 * - NestJS: @Controller('auth') + @Post('register')
 * - Java:   @RestController + @RequestMapping("/auth") + @PostMapping("/register")
 * 
 * Endpoints:
 * - POST /auth/register - Create new user account
 * - POST /auth/login    - Login and get JWT token
 * - GET  /auth/profile  - Get current user (protected)
 */
@RestController
@RequestMapping("/auth")
@RequiredArgsConstructor
public class AuthController {

    private final AuthService authService;

    /**
     * Register a new user.
     * 
     * @Valid triggers Jakarta Validation (like class-validator in NestJS)
     */
    @PostMapping("/register")
    public ResponseEntity<AuthResponse> register(@Valid @RequestBody RegisterRequest request) {
        return ResponseEntity.ok(authService.register(request));
    }

    /**
     * Login and get JWT token.
     */
    @PostMapping("/login")
    public ResponseEntity<AuthResponse> login(@Valid @RequestBody LoginRequest request) {
        return ResponseEntity.ok(authService.login(request));
    }

    /**
     * Get current user profile.
     * 
     * @AuthenticationPrincipal is like NestJS @Request() req.user
     * It injects the authenticated user from Spring Security context.
     * 
     * This endpoint is protected by SecurityConfig (requires valid JWT).
     */
    @GetMapping("/profile")
    public ResponseEntity<AuthResponse.UserDto> getProfile(@AuthenticationPrincipal User user) {
        return ResponseEntity.ok(
                AuthResponse.UserDto.builder()
                        .id(user.getId())
                        .email(user.getEmail())
                        .firstName(user.getFirstName())
                        .lastName(user.getLastName())
                        .role(user.getRole())
                        .build()
        );
    }
}
